src/mlpack/methods/ann/layer/nearest_interpolation_impl.hpp

```diff
@@ -0,0 +1,162 @@
+ * terms of the 3-clause BSD license. You should have received a copy of the
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_NEAREST_INTERPOLATION_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_NEAREST_INTERPOLATION_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "nearest_interpolation.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+template<typename InputDataType, typename OutputDataType>
+NearestInterpolation<InputDataType, OutputDataType>::
+  NearestInterpolation():
+  inRowSize(0),
+  inColSize(0),
+  outRowSize(0),
+  outColSize(0),
+  depth(0),
+  batchSize(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+NearestInterpolation<InputDataType, OutputDataType>::
+NearestInterpolation(const size_t inRowSize,
+                     const size_t inColSize,
+                     const size_t outRowSize,
+                     const size_t outColSize,
+                     const size_t depth) :
+    inRowSize(inRowSize),
+    inColSize(inColSize),
+    outRowSize(outRowSize),
+    outColSize(outColSize),
+    depth(depth),
+    batchSize(0)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void NearestInterpolation<InputDataType, OutputDataType>::Forward(
+  const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  batchSize = input.n_cols;
+  if (output.is_empty())
+    output.set_size(outRowSize * outColSize * depth, batchSize);
+  else
+  {
+    assert(output.n_rows == outRowSize * outColSize * depth);
+    assert(output.n_cols == batchSize);
+  }
+
+  assert(inRowSize >= 2);
+  assert(inColSize >= 2);
+
+  arma::cube inputAsCube(const_cast<arma::Mat<eT>&>(input).memptr(),
+      inRowSize, inColSize, depth * batchSize, false, false);
+  arma::cube outputAsCube(output.memptr(), outRowSize, outColSize,
+      depth * batchSize, false, true);
+
+  double scaleRow = (double) inRowSize / (double) outRowSize;
+  double scaleCol = (double) inColSize / (double) outColSize;
+
+  for (size_t i = 0; i < outRowSize; ++i)
+  {
+    const size_t rOrigin = std::floor(i * scaleRow);
+
+    for (size_t j = 0; j < outColSize; ++j)
+    {
+      const size_t cOrigin = std::floor(j * scaleCol);
+
+      for (size_t k = 0; k < depth * batchSize; ++k)
+      {
+        outputAsCube(i, j, k) = inputAsCube.slice(k)(
+            rOrigin, cOrigin);
+      }
+    }
+  }
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void NearestInterpolation<InputDataType, OutputDataType>::Backward(
+  const arma::Mat<eT>& /*input*/,
+  const arma::Mat<eT>& gradient,
+  arma::Mat<eT>& output)
+{
+  if (output.is_empty())
+  {
+    output.set_size(inRowSize * inColSize * depth, batchSize);
+    output.zeros();
```

Comment:
```suggestion
    output.zeros(inRowSize * inColSize * depth, batchSize);
```

Both steps combined.